libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1561,6 +1562,80 @@ string YulUtilFunctions::copyArrayToStorage(ArrayType const& _fromType, ArrayTyp
 	});
 }
 
+
+string YulUtilFunctions::copyByteArrayToStorageFunction(ArrayType const& _fromType, ArrayType const& _toType)
+{
+	solAssert(
+		*_fromType.copyForLocation(_toType.location(), _toType.isPointer()) == dynamic_cast<ReferenceType const&>(_toType),
+		""
+	);
+	solAssert(_fromType.isByteArray(), "");
+	solAssert(_toType.isByteArray(), "");
+	solUnimplementedAssert(!_fromType.dataStoredIn(DataLocation::Storage), "");
+
+	string functionName = "copy_byte_array_to_storage_from_" + _fromType.identifier() + "_to_" + _toType.identifier();
+	return m_functionCollector.createFunction(functionName, [&](){
+		Whiskers templ(R"(
+			function <functionName>(slot, src<?fromCalldata>, len</fromCalldata>) {
+				let newLen := <arrayLength>(src<?fromCalldata>, len</fromCalldata>)
+				// TODO do need to panic-check the new length?
+				let oldLen := <byteArrayLength>(sload(slot))
+
+				<?fromMemory>
+					src := add(src, 0x20)
+				</fromMemory>
+
+				// This is not needed in all branches.
+				let dstDataArea
+				if or(gt(oldLen, 31), gt(newLen, 31)) {
+					dstDataArea := <dstDataLocation>(slot)
+				}
+
+				if gt(oldLen, 31) {
+					// potentially truncate data
+					let deleteStart := add(dstDataArea, div(add(newLen, 31), 32))
+					<clearStorageRange>(deleteStart, add(src, div(add(oldLen, 31), 32)))
```

Comment:
Also do you think we can do this instead with a call to `YulUtilFunctions::resizeDynamicByteArrayFunction`?

Reply:
> Would it make sense to make a local variable with value `div(add(oldLen, 31), 32)` since it done twice?

I think it is done only once(second call is on `newLen`), I would leave as is

Reply:
> Also do you think we can do this instead with a call to `YulUtilFunctions::resizeDynamicByteArrayFunction`?

It would waste one `sload` as we did one `sload` to read the length. Other option would be to write another resize function which will accept contents of a slot instead of slot pointer... But I think it would over-complicate this